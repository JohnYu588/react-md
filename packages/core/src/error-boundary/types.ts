/**
 * @since 6.0.0
 */
export type ErrorBoundaryState =
  | {
      errored: false;
      error: null;
    }
  | {
      errored: true;
      error: Error;
    };

/**
 * @since 6.0.0
 */
export type ErrorBoundaryContext = ErrorBoundaryState & {
  reset: () => void;
};
