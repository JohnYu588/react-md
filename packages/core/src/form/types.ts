import {
  type CSSProperties,
  type HTMLAttributes,
  type InputHTMLAttributes,
  type LabelHTMLAttributes,
  type ReactNode,
} from "react";
import { type PropsWithRef } from "../types.js";

declare module "react" {
  interface CSSProperties {
    "--rmd-form-active-color"?: string;
    "--rmd-form-focus-color"?: string;
  }
}

/**
 * The supported themes for the `TextField`, `TextArea`, and `Select`
 * components.
 *
 * - "none" - display as an unstyled text field without any border or background
 *   colors.
 * - "underline" - display with only an underline that gains the form active
 *   color and animates from the left or right to the other side when the field
 *   is focused.
 * - "filled" - an extension of the `"underline"` state that will also have a
 *   slightly dark background applied.
 * - "outline" - outlines the entire text field in a border and applies the
 *   active color as box shadow when the field is focused.
 */
export type FormTheme = "none" | "underline" | "filled" | "outline";

/**
 * The direction that the underline should appear from when the theme is
 * `"underline"` or `"filled"`.
 */
export type FormUnderlineDirection = "left" | "center" | "right";

export interface FormThemeOptions {
  /**
   * The current theme type.
   *
   * @defaultValue `"outline"`
   */
  theme?: FormTheme;

  /**
   * The current underline direction.
   *
   * @defaultValue `"left"`
   */
  underlineDirection?: FormUnderlineDirection;
}

/**
 * @since 6.0.0
 */
export interface FormConfiguration extends Required<FormThemeOptions> {
  /**
   * Set this to `false` if the `$disable-uncontrolled-input-toggles` variable
   * is set to `true` in the Sass configuration.
   *
   * Since the `checked` state only changes for the radio that has been clicked,
   * the previously checked radio would also be shown as checked with no way of
   * fixing it without controlling the radio component. When this flag is
   * enabled, the checked icons and state are handled through css instead of
   * `useState`.
   *
   * @defaultValue `true`
   */
  uncontrolledToggles: boolean;
}

export interface FormComponentStates {
  /** @defaultValue `false` */
  error?: boolean;

  /** @defaultValue `false` */
  active?: boolean;

  /** @defaultValue `false` */
  disabled?: boolean;

  /** @defaultValue `false` */
  readOnly?: boolean;
}

/**
 * @since 6.0.0
 * @see https://html.spec.whatwg.org/multipage/forms.html#autofill
 * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete#values
 */
export type AutocompleteValue =
  | "off"
  | "on"
  | "name"
  | "honorific-prefix"
  | "given-name"
  | "additional-name"
  | "family-name"
  | "honorific-suffix"
  | "nickname"
  | "email"
  | "username"
  | "new-password"
  | "current-password"
  | "one-time-code"
  | "organization-title"
  | "organization"
  | "street-address"
  | "address-line1"
  | "address-line2"
  | "address-line3"
  | "address-level1"
  | "address-level2"
  | "address-level3"
  | "address-level4"
  | "country"
  | "country-name"
  | "postal-code"
  | "cc-name"
  | "cc-given-name"
  | "cc-additional-name"
  | "cc-family-name"
  | "cc-number"
  | "cc-exp"
  | "cc-exp-month"
  | "cc-exp-year"
  | "cc-csc"
  | "cc-type"
  | "transaction-currency"
  | "transaction-amount"
  | "language"
  | "bday"
  | "bday-day"
  | "bday-month"
  | "bday-year"
  | "sex"
  | "tel"
  | "tel-country-code"
  | "tel-national"
  | "tek-area-code"
  | "tel-local"
  | "tel-extension"
  | "impp"
  | "url"
  | "photo";

/**
 * @since 6.0.0
 */
export interface UserAgentAutocompleteProps {
  /**
   * Set this to enable additional autocompletion suggestions for a user for
   * different form fields. Using this prop will update the
   * {@link UserAgentAutocompleteProps.name} and {@link autoComplete} to default to
   * this value.
   *
   * @example
   * ```tsx
   * <Form>
   *   <TextField
   *     label="Enter your credit card number"
   *     autoCompleteValue="cc-number"
   *     {...creditCardProps}
   *     inputMode="number"
   *   />
   *   <TextField
   *     label="Name on card"
   *     autoCompleteValue="cc-name"
   *     {...creditCardNameProps}
   *   />
   *   <TextField
   *     label="Security code"
   *     autoCompleteValue="cc-csc"
   *     {...securityCodeProps}
   *     inputMode="number"
   *   />
   *   <Button type="submit">Submit</Button>
   * </Form>
   * ```
   *
   * @see https://html.spec.whatwg.org/multipage/forms.html#autofill
   * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete#values
   * @see {@link AutocompleteValue}
   * @see {@link autoComplete}
   * @see {@link UserAgentAutocompleteProps.name}
   */
  autoCompleteValue?: AutocompleteValue;

  /**
   * @see {@link autoCompleteValue}
   * @defaultValue `autoCompleteValue`
   */
  autoComplete?: InputHTMLAttributes<HTMLInputElement>["autoComplete"];

  /**
   * @see {@link autoCompleteValue}
   * @defaultValue `autoCompleteValue`
   */
  name?: string;
}

export interface FormMessageClassNameOptions {
  className?: string;

  /**
   * Boolean if the message should gain the error state which changes the text
   * color to `red` by default.
   *
   * @defaultValue `false`
   */
  error?: boolean;

  /**
   * The current theme for the related text field. This is really only used to
   * match the current horizontal padding of the text field.
   *
   * @defaultValue `"outline"`
   */
  theme?: FormTheme;
}

export interface FormMessageProps
  extends Omit<HTMLAttributes<HTMLDivElement>, "minLength" | "maxLength">,
    FormMessageClassNameOptions {
  /**
   * If this component is acting as a form-level error message handler, the role
   * should be updated to be `"alert"` for additional accessibility.
   *
   * Note: when creating a form-level error message handler, the messages should
   * no longer appear as the user types and instead once the user tries to
   * submit the form. Having an alert role disrupts normal screen reader
   * behavior by immediately reading changes in this element.
   *
   * @defaultValue `undefined`
   */
  role?: "alert";

  /**
   * Boolean if the children should no longer be wrapped in a `<p>` tag. This
   * should normally only be disabled if using a custom error message wrapper or
   * the counter behavior is not being used. To get correct alignments of the
   * message and counter, the `children` must be wrapped in some element and
   * cannot be plain test.
   *
   * Note: this will always be considered `true` if the `role` is set to
   * `"alert"`.
   *
   * @defaultValue `false`
   */
  disableWrap?: boolean;

  /**
   * An optional style to apply to the `<p>` tag that surrounds the `children`.
   * This will not be used if `role="alert"` or `disableWrap={true}`.
   */
  messageStyle?: CSSProperties;

  /**
   * An optional className to apply to the `<p>` tag that surrounds the
   * `children`. This will not be used if `role="alert"` or
   * `disableWrap={true}`.
   */
  messageClassName?: string;
}

/**
 * Props that are used to automatically add a counter for the remaining letters
 * available for the text field. The counter will always be created to the right
 * of the optional message.
 *
 * The counter is really a simple string of: `${length} / ${maxLength}`.
 *
 * If you need additional customization, it is recommended to create your own
 * implementation such as:
 *
 * ```tsx
 * <FormMessage>
 *   {errorMessage}
 *   <MyCounter {...props} />
 * </FormMessage>
 * ```
 *
 * Note: this should not be used alongside form-level messages.
 *
 * @since 2.9.0 Renamed from `FormMessageCounterProps` to
 * `FormMessageInputLengthCounterProps` since a `FormMessageCounter` component
 * was added
 */
export interface FormMessageInputLengthCounterProps {
  /**
   * The current length of the value in the related text field.
   */
  length: number;

  /**
   * The max length allowed for the value in the related text field.
   */
  maxLength: number;

  /**
   * An optional style to apply to the counter wrapper element.
   */
  counterStyle?: CSSProperties;

  /**
   * An optional className to apply to the counter wrapper element.
   */
  counterClassName?: string;
}

export interface FormMessageWithCounterProps
  extends FormMessageProps,
    FormMessageInputLengthCounterProps {}

/**
 * @since 6.0.0
 */
export interface FormMessageWithoutCounterProps extends FormMessageProps {
  length?: never;
  maxLength?: never;
  counterStyle?: never;
  counterClassName?: never;
}

export interface FormMessageContainerExtension {
  /**
   * If the extension doesn't actually want to render the `FormMessage`
   * component, these props are optional. It kind of eliminates the whole
   * purpose of this component though.
   */
  messageProps?: PropsWithRef<
    FormMessageProps & Partial<FormMessageInputLengthCounterProps>,
    HTMLDivElement
  >;

  /**
   * Any props (and an optional ref) to provide to the `<div>` surrounding the
   * children and `FormMessage` component.
   *
   * Note: This will not be used if the `messageProps` are not provided since
   * only the `children` will be returned without the container.
   */
  messageContainerProps?: PropsWithRef<
    HTMLAttributes<HTMLDivElement>,
    HTMLDivElement
  >;
}

/** @since 6.0.0 */
export interface LabelClassNameOptions {
  className?: string;

  /**
   * Set this to `true` to remove the `gap` style from the label.
   *
   * @see `$label-gap`
   * @defaultValue `false`
   */
  gap?: boolean;

  /**
   * Set this to `true` when the parent `TextFieldContainer` has the `dense`
   * spec enabled. This updates the floating styles to match the smaller height.
   *
   * @defaultValue `false`
   */
  dense?: boolean;

  /**
   * Set this to `true` to update the label's color to the error color.
   *
   * @see `$error-color`
   * @defaultValue `false`
   */
  error?: boolean;

  /**
   * Set this to `true` to update the label's color to the active color.
   *
   * @see `$active-color`
   * @defaultValue `false`
   */
  active?: boolean;

  /**
   * Set this to `true` if the label should gain `flex-direction: column`
   * styling.
   *
   * @defaultValue `false`
   */
  stacked?: boolean;

  /**
   * Set this to `true` to update the label's color to be the disabled color.
   *
   * @see `$disabled-color`
   * @defaultValue `false`
   */
  disabled?: boolean;

  /**
   * Set this to true when label can floating above an input, textarea, or
   * select inside of a `TextFieldContainer`.
   *
   * @defaultValue `false`
   */
  floating?: boolean;

  /**
   * Set this to true when label is currently floating above an input,
   * textarea, or selected inside of a `TextFieldContainer`.
   *
   * @see {@link active}
   * @defaultValue `active`
   */
  floatingActive?: boolean;

  /**
   * Set this to `true` to gain `flex-direction: row-reversed` styling. If the
   * {@link stacked} prop is also `true`, `flex-direction: column-reversed` will
   * be applied.
   *
   * @defaultValue `false`
   */
  reversed?: boolean;

  /**
   * @defaultValue `false`
   */
  inactive?: boolean;
}

export interface LabelProps
  extends LabelHTMLAttributes<HTMLLabelElement>,
    LabelClassNameOptions {}

/**
 * @since 6.0.0
 */
export interface ConfigurableTextFieldAddonProps
  extends HTMLAttributes<HTMLSpanElement> {
  /**
   * Boolean if the addon should be presentational only and prevent pointer
   * events.
   *
   * @defaultValue `false`
   */
  pointerEvents?: boolean;
}

/**
 * @since 6.0.0 Added support for `leftAddonProps` and
 * `rightAddonProps`.
 */
export interface TextFieldContainerOptions
  extends FormThemeOptions,
    FormComponentStates {
  /**
   * Set this to `true` to enable the dense spec which reduces the height.
   *
   * @defaultValue `false`
   */
  dense?: boolean;

  /**
   * Set this to `true` to change the style from `display: flex` to
   * `display: inline-flex`.
   *
   * @defaultValue `false`
   */
  inline?: boolean;

  /**
   * This should generally be an icon or a button that will be placed before the
   * `TextField` or `TextArea`.
   */
  leftAddon?: ReactNode;

  /**
   * Any additional props to pass to the `<span>` surrounding the {@link leftAddon}.
   *
   * @since 6.0.0
   */
  leftAddonProps?: PropsWithRef<
    ConfigurableTextFieldAddonProps,
    HTMLSpanElement
  >;

  /**
   * @see {@link TextFieldAddonProps.disabled}
   *
   * @defaultValue `false`
   */
  disableLeftAddonStyles?: boolean;

  /**
   * This should generally be an icon or a button that will be placed after the
   * `TextField` or `TextArea`.
   */
  rightAddon?: ReactNode;

  /**
   * Any additional props to pass to the `<span>` surrounding the {@link rightAddon}.
   *
   * @since 6.0.0
   */
  rightAddonProps?: PropsWithRef<
    ConfigurableTextFieldAddonProps,
    HTMLSpanElement
  >;

  /**
   * @see {@link TextFieldAddonProps.disabled}
   *
   * @defaultValue `false`
   */
  disableRightAddonStyles?: boolean;
}

export interface FormFieldOptions
  extends TextFieldContainerOptions,
    FormMessageContainerExtension {
  /**
   * An optional floating label to use with the text field. A label is generally
   * recommended for accessibility, but can be omitted if an `aria-label` or
   * `aria-labelledby` is provided.
   */
  label?: ReactNode;

  /**
   * Any additional props and/or ref that should be passed to the `<label>`
   * element when a {@link label} is provided.
   *
   * @example
   * ```tsx
   * labelProps={{
   *   ref: labelRef,
   *   style: {},
   *   className: "some-custom-class-name",
   *   onClick: (event) => {
   *     // do something
   *   }
   * }}
   * ```
   */
  labelProps?: PropsWithRef<LabelProps, HTMLLabelElement>;

  /**
   * A convenience prop to apply a custom style to a label. This is equivalent
   * to:
   *
   * ```ts
   * labelProps={{
   *   style: // some style here
   * }}
   * ```
   */
  labelStyle?: CSSProperties;

  /**
   * A convenience prop to apply a custom className to a label. This is
   * equivalent to:
   *
   * ```ts
   * labelProps={{
   *   className: "some-class-name",
   * }}
   * ```
   */
  labelClassName?: string;
}
