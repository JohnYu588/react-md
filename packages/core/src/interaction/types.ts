import type {
  DragEventHandler,
  FocusEventHandler,
  KeyboardEventHandler,
  MouseEventHandler,
  TouchEventHandler,
} from "react";

declare module "react" {
  interface CSSProperties {
    "--rmd-ripple-background-color"?: string;
    "--rmd-interaction-background-color"?: string;
    "--rmd-hover-background-color"?: string;
    "--rmd-focus-background-color"?: string;
    "--rmd-press-background-color"?: string;
    "--rmd-selected-background-color"?: string;
    "--rmd-focus-color"?: string;
    "--rmd-focus-width"?: string | number;
  }
}

/**
 * This is used to provide feedback to the user that they are interacting with
 * elements on the page. It is recommended to not set this to `"none"` unless
 * you will implement your own version.
 *
 * When this is set to `"press"`, the `background-color` for the element will
 * become slightly darker while the user:
 * - is holding the mouse down on the element
 * - holding the enter or space key on the element
 * - holding their finger on the element for touch devices
 *
 * The `background-color` will transition in and out based on the pressed state.
 *
 * When this is set to `"ripple"`, a water droplet type of animation will appear
 * from the current coordinates of the mouse or touch event within the element.
 * Keyboard events will just trigger the animation from the center of the
 * element. Once the user stops pressing the element, the animation will start
 * to fade out.
 *
 * Note: this should match the `$interaction-mode` SCSS variable.
 *
 * @defaultValue `"ripple"`
 * @remarks \@since 6.0.0
 */
export type ElementInteractionMode = "ripple" | "press" | "none";

/** @remarks \@since 6.0.0 */
export interface ElementInteractionHandlers<E extends HTMLElement> {
  onBlur: FocusEventHandler<E>;

  /**
   * The click event handler is is only used to display a ripple for
   * `<button type="submit" />` since pressing enter on form elements should
   * submit the form. This ripple is really just to help show that the form has
   * been submitted.
   */
  onClick: MouseEventHandler<E>;

  /**
   * The keydown event handler is used to either activate the `pressed` state
   * for the element or start the ripple animation when the `enter` or `space`
   * keys are pressed. It will also trigger a click event for elements that do
   * not support this natively (`<button>`, `<a>`).
   *
   * When the space key is pressed, `event.preventDefault()` will also be called
   * to prevent the page from scrolling.
   */
  onKeyDown: KeyboardEventHandler<E>;

  /**
   * The keyup event handler is used to either deactivate the `pressed` state
   * for the element or start the exit animation for the ripple if the ripple
   * was activated by the keydown event.
   */
  onKeyUp: KeyboardEventHandler<E>;

  /**
   * The mousedown event handler is used to either activate the `pressed` state
   * for the element or start the ripple animation.
   */
  onMouseDown: MouseEventHandler<E>;

  /**
   * The mouseup event handler is used to either deactivate the `pressed` state
   * for the element or start the exit animation for the ripple if the ripple
   * was activated by the mousedown event.
   */
  onMouseUp: MouseEventHandler<E>;

  /**
   * The mouseleave event handler will remove all ripples and prevent any other
   * interactions if the current {@link UserInteractionMode} is `"mouse"`.
   *
   * The ripples have to be cancelled since the user might release the mouse
   * outside of the element which would never trigger the `onMouseUp` flow.
   */
  onMouseLeave: MouseEventHandler<E>;

  /**
   * The dragstart event handler will remove the ripples and prevent any other
   * interactions.
   */
  onDragStart: DragEventHandler<E>;

  /**
   * The touchstart event handler is used to either activate the `pressed` state
   * for the element or start the ripple animation.
   */
  onTouchStart: TouchEventHandler<E>;

  /**
   * The touchend event handler is used to either deactivate the `pressed`
   * state for the element or start the exit animation for the ripple if the
   * ripple was activated by the touchstart event.
   */
  onTouchEnd: TouchEventHandler<E>;

  /**
   * The touchmove event handler will remove all ripples and prevent any other
   * interactions if the current {@link UserInteractionMode} is `"touch"`.
   */
  onTouchMove: TouchEventHandler<E>;
}

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export interface RippleStyle {
  readonly left: number;
  readonly top: number;
  readonly height: number;
  readonly width: number;
}

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export interface RippleState {
  readonly style: RippleStyle;
  readonly exiting: boolean;
  readonly entered: boolean;
  readonly startTime: number;
}

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export type RippleStateList = readonly RippleState[];

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export interface RippleTransitionCallbacks {
  onEntered(ripple: RippleState): void;
  onExited(ripple: RippleState): void;
}

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export interface ProvidedRippleContainerProps
  extends RippleTransitionCallbacks {
  ripples: RippleStateList;
}

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export interface ElementInteractionState {
  pressed: boolean;
  ripples: RippleStateList;
}
