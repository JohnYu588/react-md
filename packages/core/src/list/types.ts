import { type HTMLAttributes, type ReactNode } from "react";

import { type PropsWithRef } from "../types.js";

/**
 * The height to apply to the list item.
 *
 * Conversions:
 *
 * - height !== undefined && height !== "auto" -&gt; height
 * - secondaryText or left/right addon is media/media-large  -&gt; "extra-large"
 * - left/right addon is avatar -&gt; "large"
 * - left/right addon is icon -&gt; "medium"
 * - no addons and no secondary text -&gt; "normal"
 */
export type ListItemHeight =
  | "auto"
  | "normal"
  | "medium"
  | "large"
  | "extra-large";

export type ListItemAddonPosition = "top" | "middle" | "bottom";
export type ListItemAddonType = "icon" | "avatar" | "media" | "large-media";

/** @since 6.0.0 */
export interface ListItemChildrenAddonProps {
  /**
   * An optional addon to display to the left of the `primaryText` or
   * `children` and should be used with the `leftAddonType` prop to adjust
   * spacing.
   */
  leftAddon?: ReactNode;

  /**
   * The type of the addon that appears to the left of the `primaryText` or
   * `children`.
   *
   * @defaultValue `"icon"`
   */
  leftAddonType?: ListItemAddonType;

  /**
   * The vertical position the left icon, avatar, media, or large media
   * should be placed.
   *
   * @defaultValue `"middle"`
   */
  leftAddonPosition?: ListItemAddonPosition;

  /**
   * @since 6.0.0
   */
  leftAddonClassName?: string;

  /**
   * Set this to `true` if your addon does not accept a `className` prop to
   * apply the correct styling.
   *
   * @defaultValue `leftAddonType === "media" || leftAddonType === "large-media"`
   * @since 6.0.0
   */
  leftAddonForceWrap?: boolean;

  /**
   * Set this to `true` to disable the additional spacing applied to the addons
   * that appear before the `children` and the normal `--rmd-text-icon-spacing`
   * value is used instead.
   *
   * @defaultValue `false`
   */
  disableLeftAddonSpacing?: boolean;

  /**
   * An optional addon to display to the right of the `primaryText` or
   * `children` and should be used with the `rightAddonType` prop to adjust
   * spacing.
   */
  rightAddon?: ReactNode;

  /**
   * The type of the addon that appears to the right of the `primaryText` or
   * `children`.
   *
   * @defaultValue `"icon"`
   */
  rightAddonType?: ListItemAddonType;

  /**
   * The vertical position the right icon, avatar, media, or large media
   * should be placed.
   *
   * @defaultValue `"middle"`
   */
  rightAddonPosition?: ListItemAddonPosition;

  /**
   * @since 6.0.0
   */
  rightAddonClassName?: string;

  /**
   * Set this to `true` if your addon does not accept a `className` prop to
   * apply the correct styling.
   *
   * @defaultValue `rightAddonType === "media" || rightAddonType === "large-media"`
   * @since 6.0.0
   */
  rightAddonForceWrap?: boolean;

  /**
   * The media items are centered by default using:
   * ```scss
   * align-items: center;
   * display: flex;
   * justify-content: center;
   * ```
   *
   * When this is set to `true`, the flex styles will not be applied.
   *
   * @defaultValue `false`
   * @since 6.0.0
   */
  disableLeftAddonCenteredMedia?: boolean;

  /**
   * The media items are centered by default using:
   * ```scss
   * align-items: center;
   * display: flex;
   * justify-content: center;
   * ```
   *
   * When this is set to `true`, the flex styles will not be applied.
   *
   * @defaultValue `false`
   * @since 6.0.0
   */
  disableRightAddonCenteredMedia?: boolean;
}

/**
 * Note: This interface was added since there are components that replace the
 * {@link ListItemChildrenAddonProps} with renamed versions but should still
 * support everything for rendering text.
 *
 * @since 6.0.0
 */
export interface ListItemChildrenTextProps {
  /**
   * The main content to display that defaults to being wrapped in the
   * `ListItemText` component. Enable the {@link disableTextChildren} prop to
   * render without the additional `ListItemText` wrapper.
   */
  children?: ReactNode;

  /**
   * Any additional props that should be passed to the `<span>` surrounding
   * the {@link children}, {@link primaryText}, and {@link secondaryText}.
   *
   * @since 6.0.0
   */
  textProps?: PropsWithRef<HTMLAttributes<HTMLSpanElement>>;

  /**
   * An optional className to apply to the `<span>` that surrounds the
   * `primaryText` and optionally `secondaryText` within the list item.
   */
  textClassName?: string;

  /**
   * An optional className to apply to the `<span>` that surrounds the
   * `secondaryText` within the list item.
   */
  secondaryTextClassName?: string;

  /**
   * Set this to `true` if the {@link children} should not automatically be
   * wrapped in the `ListItemText` component.
   *
   * @since 6.0.0 Renamed from `textChildren` since it was defaulted on.
   * @defaultValue `false`
   */
  disableTextChildren?: boolean;

  /**
   * An optional element that should be rendered as the `primaryText` within the
   * list item. It is most likely easier to use the `children` prop instead, but
   * this allows you to create more complex components with the `ListItem` since
   * you can provided `children` and have the styles for the `primaryText` still
   * applied. By default, this will only allow one line of text and add ellipsis
   * for any text overflow.
   */
  primaryText?: ReactNode;

  /**
   * An optional element that should be rendered as the `secondaryText` within
   * the list item. By default, this will only span one line and add ellipsis
   * for overflow.
   */
  secondaryText?: ReactNode;

  /**
   * Any additional props that should be passed to the `<span>` surrounding
   * the {@link secondaryText}.
   *
   * @since 6.0.0
   */
  secondaryTextProps?: PropsWithRef<HTMLAttributes<HTMLSpanElement>>;

  /**
   * Set this to `true` to allow the {@link secondaryText} to span multiple
   * lines and use the CSS clamp after two lines of text are visible. The number
   * of lines can be configured by `core.$list-item-multiline-clamp` or
   * `@include core.list-set-var(item-multiline-clamp, $new-value)`.
   *
   * @see {@link https://developer.mozilla.org/en-US/docs/Web/CSS/-webkit-line-clamp}
   * @defaultValue `false`
   */
  multiline?: boolean;
}

/**
 * @since 6.0.0 Extends the `ListItemChildrenTextProps` and
 * `ListItemChildrenAddonProps` interfaces.
 * @since 6.0.0 Added `textProps` and `secondaryTextClassName`
 * @since 6.0.0 Renamed `textChildren` to `disableTextChildren`
 */
export interface ListItemChildrenProps
  extends ListItemChildrenTextProps,
    ListItemChildrenAddonProps {}
