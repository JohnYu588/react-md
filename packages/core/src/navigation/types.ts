import {
  type AnchorHTMLAttributes,
  type ForwardRefExoticComponent,
  type HTMLAttributes,
  type ReactNode,
} from "react";

import { type DividerProps } from "../divider/Divider.js";
import { type ListSubheaderProps } from "../list/ListSubheader.js";
import {
  type TooltipOptions,
  type TooltippedElementEventHandlers,
} from "../tooltip/useTooltip.js";

/**
 * Used to add a `Divider` to the navigation list.
 * @example Output
 * ```tsx
 * const { type, ...props } = item;
 * return <Divider {...props} />;
 * ```
 *
 * @since 6.0.0
 */
export interface NavigationItemDivider extends DividerProps {
  type: "divider";
}

/**
 * Used to add a `ListSubheader` to the navigation list.
 * @example Output
 * ```tsx
 * const { type, ...props } = item;
 * return <ListSubheader {...props} />;
 * ```
 *
 * @since 6.0.0
 */
export interface NavigationItemSubheader extends ListSubheaderProps {
  type: "subheader";
  children: ReactNode;
}

/**
 * This is used to either create a route group which update all child routes to
 * be prefixed with the optional `href`.
 *
 * @example Route Group Output
 * ```tsx
 * const nextParents = [item, ...parents];
 * const id = getNavigationGroupId(item, nextParents);
 * return (
 *   <CollapsibleNavGroup
 *     depth={nextParents.length}
 *     collapsed={!data?.expandedItems.has(id)}
 *     buttonChildren={item.children}
 *     toggleCollapsed={() => {
 *       data?.toggleExpandedItem(id);
 *     }}
 *   >
 *     {children}
 *   </CollapsibleNavGroup>
 * );
 * ```
 *
 * Note: `children` are any child `items` that have been rendered.
 *
 * @since 6.0.0
 */
export interface NavigationItemGroup {
  type: "group";

  /**
   * An optional unique id to identify the group used for handling the expanded
   * state when the `href` can't be used.
   */
  id?: string;

  /**
   * An optional href to prefix all child routes with and can be used instead of
   * the {@link id} for handling the expanded state.
   *
   * @example
   * ```ts
   * {
   *   type: "group",
   *   href: "/some-path",
   *   children: "Some Name",
   *   items: [
   *     {
   *       type: "route",
   *       href: "/page-1",
   *       children: "Page 1",
   *     },
   *     {
   *       type: "route",
   *       href: "/page-2",
   *       children: "Page 2",
   *     },
   *   ],
   * }
   * ```
   *
   * Would result in:
   *
   * - Page 1 - `/some-path/page-1`
   * - Page 2 - `/some-path/page-2`
   */
  href?: string;
  children: ReactNode;
  items: readonly NavigationItem[];
}

/**
 * @since 6.0.0
 */
export interface NavigationItemRoute {
  type: "route";
  href: string;
  children: ReactNode;
  active?: boolean;
  beforeAddon?: ReactNode;
  afterAddon?: ReactNode;
}

/**
 * @since 6.0.0
 */
export type NavigationItem =
  | NavigationItemDivider
  | NavigationItemSubheader
  | NavigationItemGroup
  | NavigationItemRoute;

/**
 * @since 6.0.0
 */
export type CustomNavigationLinkComponent = ForwardRefExoticComponent<
  AnchorHTMLAttributes<HTMLAnchorElement> & { href: string }
>;

/**
 * @since 6.0.0
 */
export type NavigationLinkComponent = "a" | CustomNavigationLinkComponent;

/**
 * @since 6.0.0
 */
export interface NavigationExpansion {
  expandedItems: ReadonlySet<string>;
  toggleExpandedItem: (id: string) => void;
}

/**
 * @since 6.0.0
 */
export interface NavigationRenderData extends NavigationExpansion {
  /**
   * This is used to set the `active` state on a `NavItemLink`
   */
  pathname: string;
  linkComponent: NavigationLinkComponent;
}

/**
 * @since 6.0.0
 */
export interface NavItemContentProps {
  /**
   * Any additional props to provide the `<span>` element that wraps the
   * children such as `style`, `className`, and `ref`.
   */
  spanProps?: HTMLAttributes<HTMLSpanElement>;

  /**
   * An optional addon to render before the `children` and will not be wrapped
   * in the `span` that adds ellipsis overflow.
   */
  beforeAddon?: ReactNode;

  /**
   * An optional addon to render after the `children` and will not be wrapped
   * in the `span` that adds ellipsis overflow.
   */
  afterAddon?: ReactNode;

  /**
   * This should contain accessible text for the page that this will link to and
   * will automatically be truncated with ellipsis if it is too large. A tooltip
   * will also appear only when it has been truncated.
   */
  children: ReactNode;

  /**
   * This most likely will never need to be used, but it's a way to override any
   * tooltip options. The default behavior will position the tooltip to the
   * right and only appear if the `children` are overflown.
   *
   * @example Customize
   * ```tsx
   * <NavItemLink
   *   {...props}
   *   tooltipOptions={{
   *     overflowOnly: false,
   *     defaultPosition: "above",
   *     vhMargin: "1rem",
   *     vwMargin: "1rem",
   *   }}
   * >
   * ```
   */
  tooltipOptions?: Omit<
    TooltipOptions<HTMLAnchorElement>,
    keyof TooltippedElementEventHandlers<HTMLAnchorElement>
  >;
}
