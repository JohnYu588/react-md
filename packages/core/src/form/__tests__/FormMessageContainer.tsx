import { describe, expect, it } from "@jest/globals";
import { createRef } from "react";

import { render, screen } from "../../test-utils/index.js";
import { type PropsWithRef } from "../../types.js";
import {
  FormMessageContainer,
  type FormMessageContainerProps,
} from "../FormMessageContainer.js";

describe("FormMessageContainer", () => {
  it("should just render the children if messageProps are not defined", () => {
    const ref = createRef<HTMLDivElement>();
    const props = {
      "data-testid": "container",
      ref,
      children: <span data-testid="children">Children</span>,
    } as const;
    render(<FormMessageContainer {...props} />);
    expect(() => screen.getByTestId("container")).toThrow();
    expect(() => screen.getByTestId("children")).not.toThrow();
    expect(ref.current).toBe(null);
  });

  it("should apply the correct styling, HTML attributes, and allow a ref", () => {
    const ref = createRef<HTMLDivElement>();
    const props: PropsWithRef<FormMessageContainerProps> = {
      "data-testid": "container",
      ref,
      children: <span>Children</span>,
      messageProps: {
        children: "Message!",
      },
    };
    const { rerender } = render(<FormMessageContainer {...props} />);

    const container = screen.getByTestId("container");
    expect(ref.current).toBeInstanceOf(HTMLDivElement);
    expect(ref.current).toBe(container);
    expect(container).toMatchSnapshot();

    rerender(
      <FormMessageContainer
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(container).toMatchSnapshot();

    rerender(
      <FormMessageContainer
        {...props}
        inline
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(container).toMatchSnapshot();
  });
});
