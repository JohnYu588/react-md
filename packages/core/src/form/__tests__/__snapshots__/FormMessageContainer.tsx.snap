// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormMessageContainer should apply the correct styling, HTML attributes, and allow a ref 1`] = `
<div
  class="rmd-form-message-container"
  data-testid="container"
>
  <span>
    Children
  </span>
  <div
    aria-live="polite"
    class="rmd-form-message rmd-form-message--outline rmd-text-secondary-color"
    id="form-message-:r0:"
  >
    <p
      class="rmd-form-message__message"
      id="form-message-:r0:-message"
    >
      Message!
    </p>
  </div>
</div>
`;

exports[`FormMessageContainer should apply the correct styling, HTML attributes, and allow a ref 2`] = `
<div
  class="rmd-form-message-container custom-class-name"
  data-testid="container"
  style="color: white;"
>
  <span>
    Children
  </span>
  <div
    aria-live="polite"
    class="rmd-form-message rmd-form-message--outline rmd-text-secondary-color"
    id="form-message-:r0:"
  >
    <p
      class="rmd-form-message__message"
      id="form-message-:r0:-message"
    >
      Message!
    </p>
  </div>
</div>
`;

exports[`FormMessageContainer should apply the correct styling, HTML attributes, and allow a ref 3`] = `
<div
  class="rmd-form-message-container rmd-form-message-container--inline custom-class-name"
  data-testid="container"
  style="color: white;"
>
  <span>
    Children
  </span>
  <div
    aria-live="polite"
    class="rmd-form-message rmd-form-message--outline rmd-text-secondary-color"
    id="form-message-:r0:"
  >
    <p
      class="rmd-form-message__message"
      id="form-message-:r0:-message"
    >
      Message!
    </p>
  </div>
</div>
`;
